DataScience/experimenter.py

```diff
@@ -0,0 +1,202 @@
+                full_command = "{0} --ignore {1}".format(full_command, ignored_namespace)
+        if interaction_list != "":
+            for interaction in interaction_list:
+                full_command = "{0} -q {1}".format(full_command, interaction)
+        if regularization != "":
+            full_command = "{0} --l1 {1}".format(full_command, regularization)
+        self.full_command = full_command
+        self.loss = None
+        
+def result_writer(queue):
+    experiment_file = open("experiments.tsv", "a")
+    # Read from queue and write results to file until done
+    while True:
+        msg = queue.get()
+        if (msg == 'complete'):
+            break
+        command = msg
+        line = "{0:7f}\t{1}\t{2:7f}\t{3}\t{4}\t{5}\t{6}\t{7}\t{8}".format(float(command.loss), \
+            command.base, command.learning_rate, command.cb_type, str(command.marginal_list), \
+            str(command.ignore_list), str(command.interaction_list), str(command.regularization), str(datetime.now()))
+        experiment_file.write(line + "\n")
+        experiment_file.flush()
+        
+    experiment_file.close()
+    
+def run_experiment(command, timeout=1000):
+    while True:
+        try:
+            results = check_output(command.full_command.split(' '), stderr=STDOUT, timeout=timeout).decode("utf-8")
+            break
+        except TimeoutExpired:
+            print("Timeout expired for command {0}".format(command.full_command))
+            
+    m = re.search('average loss = (.+)\n', str(results))
+    loss = m.group(1)
+    command.loss = float(loss)
+    result_queue.put(command)
+    return command
+    
+def run_experiment_set(command_list):
+    # Run the experiments in parallel using 5 processes
+    p = Pool(5)
+    results = p.map(run_experiment, command_list)
+    results.sort(key=lambda result: result.loss)
+    return results
+    
+if __name__ == '__main__':
+    # Identify namespaces and detect marginal features
+    if len(sys.argv) < 2:
+        print("Usage: python experimenter.py {file_name}. Where file_name is the merged Decision Service logs in JSON format")
+        sys.exit()
+    data = open(sys.argv[1], 'r', encoding="utf8")
+    counter = 0
+    shared_features = []
+    action_features = []
+    marginal_features = []
+    for line in data:
+        counter += 1
+        event = json.loads(line)
+        for feature in event.keys():
+            if feature[0] != '_' and feature[0] not in shared_features:
+                shared_features.append(feature[0])
+        action_set = event['_multi']
+        for action in action_set:
+            for feature in action.keys():
+                if feature[0] != '_' and feature[0] not in action_features:
+                    action_features.append(feature[0])
+                    if action[feature].get('constant', 0) == 1 and 'id' in action[feature]:
+                        marginal_features.append(feature[0])
+        if counter >= 50:
+            break
```

Comment:
We assume schema is fixed so we only need to read a few lines to capture all the features. I added a comment.